File: B26.cpp
/*
=== B26 =======================================
B26. 
Dots naturāls skaitlis n. Noskaidrot, vai doto skaitli var izteikt kā trīs naturālu skaitļu kvadrātu summu. 
Izdrukāt visus  šādus naturālu skaitļu trijniekus. 
Risinājumā izmantot funkciju, kas aprēķina trīs skaitļu kvadrātu summu.
*/

#include <iostream>
using namespace std;

int sq(int x){              // skaitļa kvadarāts
    return x * x;
}

int three_n_sqd_sum(int i, int j, int k){   // 3 skaitļa kvadarātu summa
    return sq(i) + sq(j) + sq(k);
}

void find_combinations_wo_formula(int x){
    int cnt = 0;
    for (int i=1; sq(i)<=x; i++){
        for (int j=1; sq(j)<=x-sq(i); j++){
            for (int k=1; sq(k)<=x-(sq(i)+sq(j)); k++){
                //if (sq(k)==x-(sq(i)+sq(j))){ // alternativa rezultātu meklēšana
                if (three_n_sqd_sum(i, j, k) == x){ //Izmantojot formulu 
                    cout << "Numbers " << i << ", " << j << ", " << k << ", " << "sum: " << x << "\n";
                    cnt++;
                }
            }  
        }  
    }
    if (cnt == 0) cout << "This number has no valid combinations." << "\n";
}

int main()
{
    char opt, yn;
    do
    {
        int x;
        cout << "Please enter a number: ";
        cin >> x;
        if (cin.fail())     // x nav int
        {
            cin.clear();
            cout << "That was not a valid number!\n ";
        }
        else if (x <= 0) cout << "That was not a valid number!\n ";
        else                // x ir int
            find_combinations_wo_formula(x);

        cout << "Try again? (y/n): "; 
        do                  // Try again loop sagaida tikai vēr†ības (y,n)
        {
            cin >> opt; yn = opt;
        } while (yn != 'y' && yn != 'n');
        //opt='n';

    } while (opt == 'y');
    return 0;
}
/*
    "test_input,expected",
    [
        (234, "numbers 1,8,13  squared: 1,64,169 sum: 234"),
        (0, "That was not a valid number."),
        (k, "That was not a valid number."),
        (10, "This number has no valid combinations.""),
        (-10, "That was not a valid number."),
    ],
*/

File: B26.py
'''
=== B26 =======================================
B26. 
Dots naturāls skaitlis n. Noskaidrot, vai doto skaitli var izteikt kā trīs naturālu skaitļu kvadrātu summu. 
Izdrukāt visus  šādus naturālu skaitļu trijniekus. 
Risinājumā izmantot funkciju, kas aprēķina trīs skaitļu kvadrātu summu.
'''

def three_n_squared_sum(i,j,k): # 3 skaitļa kvadarātu summa
    sum = i**2 + j**2 + k**2
    return sum

def return_all_combinations(x): # 
    max = int(x**(1/2))                                 # max ir summas kvadråtsakne
    cnt = 0
    for i in range(1, max + 1):
        for j in range(1, x - i**2 + 1):
            for k in range(1, x - (i**2 + j**2) + 1):       # Var iztikt bez 3 for loop, bet prasība ir izmantot minēto funkciju
                if three_n_squared_sum(i,j,k) == x:
                    print(f'Numbers {i},{j},{k} sum: {three_n_squared_sum(i,j,k)}')
                    cnt += 1
    if cnt == 0:
        print("This number has no valid combinations.")


def menu():
    opt = "y"
    while opt == "y":
        try:
            x = int(input("Please enter a number: "))   # x ir int
            if x > 0:                                   # nav 0 vai negatīvs
                return_all_combinations(x)              # Aloritma izsaukums
            else:
                print("That was not a valid number.")   
            opt = input("Try again? (y/n)\n")           # Try again loop sagaida tikai vēr†ības (y,n)
        except ValueError:                              # x nav int
            print("That was not a valid number.")
            opt = input("Try again? (y/n)\n")           # Try again loop sagaida tikai vēr†ības (y,n)


if __name__ == "__main__":
    menu()
"""
    "test_input,expected",
    [
        (3, "Numbers 1, 1, 1, sum: 3"),
        (234, "Numbers 1,8,13  squared: 1,64,169 sum: 234"),
        (0, "That was not a valid number."),
        (k, "That was not a valid number."),
        (10, "This number has no valid combinations.""),
        (-10, "That was not a valid number."),
    ],
"""